lib/src/cli.dart

```diff
@@ -0,0 +1,187 @@
+    ..addFlag('help',
+        abbr: 'h', negatable: false, help: 'Show usage information.')
+    ..addFlag('stats',
+        abbr: 's', negatable: false, help: 'Show lint statistics.')
+    ..addFlag('benchmark', negatable: false, help: 'Show lint benchmarks.')
+    ..addFlag('visit-transitive-closure',
+        help: 'Visit the transitive closure of imported/exported libraries.')
+    ..addFlag('quiet', abbr: 'q', help: "Don't show individual lint errors.")
+    ..addFlag('machine',
+        help: 'Print results in a format suitable for parsing.',
+        defaultsTo: false,
+        negatable: false)
+    ..addFlag('strong', help: 'Use strong-mode analyzer.')
+    ..addOption('config', abbr: 'c', help: 'Use configuration from this file.')
+    ..addOption('dart-sdk', help: 'Custom path to a Dart SDK.')
+    ..addMultiOption('rules',
+        help: 'A list of lint rules to run. For example: '
+            'avoid_as,annotate_overrides')
+    ..addOption('packages',
+        help: 'Path to the package resolution configuration file, which\n'
+            'supplies a mapping of package names to paths.  This option\n'
+            'cannot be used with --package-root.')
+    ..addOption('package-root',
+        abbr: 'p', help: 'Custom package root. (Discouraged.)');
+
+  ArgResults options;
+  try {
+    options = parser.parse(args);
+  } on FormatException catch (err) {
+    printUsage(parser, errorSink, err.message);
+    exitCode = unableToProcessExitCode;
+    return;
+  }
+
+  if (options['help']) {
+    printUsage(parser, outSink);
+    return;
+  }
+
+  if (options.rest.isEmpty) {
+    printUsage(parser, errorSink,
+        'Please provide at least one file or directory to lint.');
+    exitCode = unableToProcessExitCode;
+    return;
+  }
+
+  var lintOptions = initialLintOptions;
+
+  var configFile = options['config'];
+  if (configFile != null) {
+    var config = new LintConfig.parse(readFile(configFile));
+    lintOptions.configure(config);
+  }
+
+  var lints = options['rules'];
+  if (lints != null && !lints.isEmpty) {
+    var rules = <LintRule>[];
+    for (var lint in lints) {
+      var rule = Registry.ruleRegistry[lint];
+      if (rule == null) {
+        errorSink.write('Unrecognized lint rule: $lint');
+        exit(unableToProcessExitCode);
+      }
+      rules.add(rule);
+    }
+
+    lintOptions.enabledLints = rules;
+  }
+
+  var customSdk = options['dart-sdk'];
+  if (customSdk != null) {
+    lintOptions.dartSdkPath = customSdk;
+  }
+
+  var strongMode = options['strong'];
+  if (strongMode != null) lintOptions.strongMode = strongMode;
+
+  var customPackageRoot = options['package-root'];
+  if (customPackageRoot != null) {
+    lintOptions.packageRootPath = customPackageRoot;
+  }
+
+  var packageConfigFile = options['packages'];
+
+  if (customPackageRoot != null && packageConfigFile != null) {
+    errorSink.write("Cannot specify both '--package-root' and '--packages'.");
+    exitCode = unableToProcessExitCode;
```

Comment:
pro tip: return the exit code from this method – set it in `bin`. Makes testing MUCH easier.

Reply:
Great point.  That said, I think these cases are pretty well covered in our integration tests.  Are you making a general point or do you see cases we're missing?